optuna/integration/botorch.py

```diff
@@ -0,0 +1,550 @@
+            data, the objectives, the constraints, the search space bounds and return the next
+            candidates. The arguments are of type ``torch.Tensor``. The return value must be a
+            ``torch.Tensor``. However, if ``constraints_func`` is omitted, constraints will be
+            :obj:`None`.
+
+            If omitted, is determined automatically based on the number of objectives. If the
+            number of objectives is one, Quasi MC-based batch Expected Improvement (qEI) is used.
+            If the number of objectives is larger than one but smaller than four, Quasi MC-based
+            batch Expected Hypervolume Improvement (qEHVI) is used. Otherwise, for larger number
+            of objectives, the faster Quasi MC-based extended ParEGO (qParEGO) is used.
+
+            The function should assume *maximization* of the objective.
+
+            .. seealso::
+                See :func:`optuna.integration.botorch.qei_candidates_func` for an example.
+        constraints_func:
+            An optional function that computes the objective constraints. It must take a
+            :class:`~optuna.multi_objective.study.MultiObjectiveStudy`, a
+            :class:`~optuna.multi_objective.trial.FrozenMultiObjectiveTrial` and return the
+            constraints. The return value must be a sequence of :obj:`float` s. A value strictly
+            larger than 0 means that a constraints is violated. A value equal to or smaller than 0
+            is considered feasible.
+
+            If omitted, no constraints will be passed to ``candidates_func`` nor taken into
+            account during suggestion if ``candidates_func`` is omitted.
+
+            .. note::
+                ``constraints_func`` is called once per trial for each trial on each worker during
+                distributed optimization. Therefore, during distributed optimization, this
+                function should be deterministic to ensure that all workers hold the same values.
+        n_startup_trials:
+            Number of initial trials, that is the number of trials to resort to independent
+            sampling.
+        independent_sampler:
+            An independent sampler to use for the initial trials and for parameters that are
+            conditional.
+    """
+
+    def __init__(
+        self,
+        candidates_func: Callable[
+            [
+                "torch.Tensor",
+                "torch.Tensor",
+                Optional["torch.Tensor"],
+                "torch.Tensor",
+            ],
+            "torch.Tensor",
+        ] = None,
+        constraints_func: Optional[
+            Callable[
+                [
+                    "MultiObjectiveStudy",
+                    "FrozenMultiObjectiveTrial",
+                ],
+                Sequence[float],
+            ]
+        ] = None,
+        n_startup_trials: int = 10,
+        independent_sampler: Optional[BaseMultiObjectiveSampler] = None,
```

Comment:
For future extensibility, how about making these arguments "keyword only"? In the future, we may utilize more features of BoTorch and add more arguments, such as batch size for batch optimization, `fidelity_func` for multi-fidelity optimization, etc.

Reply:
Makes sense, fixed.